bug: storage webhook fails with 500 when artifact has no changes, blocking run completion

## Description

When an agent run completes without modifying an artifact, the storage webhook fails with 500 error instead of correctly creating an empty version. This blocks the `vm0_result` event from being sent, causing the CLI to timeout.

## Root Cause

1. Agent completes successfully but doesn't modify the `user-files` artifact
2. Storage webhook receives empty/unchanged content to upload
3. Webhook fails with ENOENT error when scanning non-existent `extracted` directory
4. Webhook returns 500 instead of handling this case gracefully
5. No `vm0_result` or `vm0_error` event is sent to CLI
6. CLI times out after 120 seconds waiting for events

## Expected Behavior

1. Storage webhook should handle empty/unchanged artifacts correctly
2. Should create a new version even if content is empty or unchanged
3. If server encounters an error, should emit `vm0_error` event to CLI
4. CLI should receive either `vm0_result` (success) or `vm0_error` (failure) and exit promptly

## Actual Behavior

- Webhook crashes with 500 error
- No completion event sent to CLI
- CLI waits 120 seconds then times out

## Reproduction Steps

1. Run an agent with an artifact that won't be modified:
   ```bash
   vm0 run demo --artifact-name user-files "my question is 1+1=?"
   ```
2. Agent answers the question without modifying any files
3. Observe: agent shows success, then run times out

## Error Logs

**Storage webhook error:**
```
2025-11-29 14:39:00.791 [error] [ERROR] [webhook:storages] Error: ENOENT: no such file or directory, scandir '/tmp/vm0-storage-webhook-1764427140780/extracted'
```

## Required Fixes

1. **Storage webhook**: Handle case where artifact has no changes (empty extracted directory or no directory at all)
2. **Error handling**: When webhook fails, emit `vm0_error` event to CLI so it doesn't hang

## Impact

Any agent run with artifacts that don't get modified will timeout, even when the agent execution succeeds.

--- a/e2e/tests/02-commands/t09-vm0-artifact-empty.bats
+++ b/e2e/tests/02-commands/t09-vm0-artifact-empty.bats
@@ -1,0 +1,79 @@
+#!/usr/bin/env bats
+
+# Test VM0 agent run with empty/unchanged artifact
+# Verifies that storage webhook correctly handles empty zip uploads
+#
+# This test covers issue #305: storage webhook fails with 500 when artifact has no changes
+
+load '../../helpers/setup'
+
+setup() {
+    export TEST_ARTIFACT_DIR="$(mktemp -d)"
+    export ARTIFACT_NAME="e2e-empty-artifact-$(date +%s)"
+    export TEST_CONFIG="${TEST_ROOT}/fixtures/configs/vm0-standard.yaml"
+}
+
+teardown() {
+    if [ -n "$TEST_ARTIFACT_DIR" ] && [ -d "$TEST_ARTIFACT_DIR" ]; then
+        rm -rf "$TEST_ARTIFACT_DIR"
+    fi
+}
+
+@test "Build VM0 empty artifact test agent configuration" {
+    run $CLI_COMMAND build "$TEST_CONFIG"
+    assert_success
+    assert_output --partial "vm0-standard"
+}
+
+@test "VM0 run with empty artifact completes successfully" {
+    # Create empty artifact (no files)
+    mkdir -p "$TEST_ARTIFACT_DIR/$ARTIFACT_NAME"
+    cd "$TEST_ARTIFACT_DIR/$ARTIFACT_NAME"
+    $CLI_COMMAND artifact init >/dev/null
+
+    # Push empty artifact
+    run $CLI_COMMAND artifact push
+    assert_success
+
+    # Run agent with operation that doesn't create any files
+    # This tests the storage webhook handling of empty zip uploads
+    run $CLI_COMMAND run vm0-standard \
+        --artifact-name "$ARTIFACT_NAME" \
+        --timeout 120 \
+        "echo 'hello world'"
+
+    assert_success
+
+    # Verify run completes properly with checkpoint
+    assert_output --partial "[result]"
+    assert_output --partial "Checkpoint:"
+}
+
+@test "VM0 run with unchanged artifact completes successfully" {
+    # Create artifact with files
+    mkdir -p "$TEST_ARTIFACT_DIR/$ARTIFACT_NAME"
+    cd "$TEST_ARTIFACT_DIR/$ARTIFACT_NAME"
+    $CLI_COMMAND artifact init >/dev/null
+
+    echo "existing content" > data.txt
+    mkdir -p subdir
+    echo "nested file" > subdir/nested.txt
+
+    run $CLI_COMMAND artifact push
+    assert_success
+
+    # Run agent that only reads files (no modifications)
+    # The storage webhook should handle unchanged artifact content correctly
+    run $CLI_COMMAND run vm0-standard \
+        --artifact-name "$ARTIFACT_NAME" \
+        --timeout 120 \
+        "cat data.txt && cat subdir/nested.txt"
+
+    assert_success
+
+    # Verify run completes properly with checkpoint
+    assert_output --partial "[result]"
+    assert_output --partial "Checkpoint:"
+    assert_output --partial "existing content"
+    assert_output --partial "nested file"
+}

--- a/turbo/apps/web/app/api/webhooks/agent/storages/route.ts
+++ b/turbo/apps/web/app/api/webhooks/agent/storages/route.ts
@@ -113,6 +113,8 @@ export async function POST(request: NextRequest) {
     // Extract zip file
     const zip = new AdmZip(zipPath);
     const extractPath = path.join(tempDir, "extracted");
+    // Ensure extract directory exists before extraction (empty zips don't create it)
+    await fs.promises.mkdir(extractPath, { recursive: true });
     zip.extractAllTo(extractPath, true);
 
     log.debug(`Extracted zip to ${extractPath}`);

